genlm/control/potential/built_in/canonical.py

```
@@ -0,0 +1,453 @@
+
+            conflict_mask = vector_k < VERYLARGE
+            conflict_mask &= vector_k <= self.vector_rp
+            conflict_mask &= vector_k < lp
+            mask[self.indices[conflict_mask][:, 0]] = False
+
+        return mask
+
+    @classmethod
+    def from_llm(cls, llm):
+        assert isinstance(llm, PromptedLLM)
+        tokenizer = llm.model.tokenizer
+        _decode, _ = decode_vocab(tokenizer)  # _decode is byte_vocab (list[bytes])
+
+        # check for duplicate byte sequences in _decode
+        byte_to_ids = defaultdict(list)
+        for token_id, byte_sequence in enumerate(_decode):
+            if (
+                byte_sequence is not None
+            ):  # Ignore potential None entries if decode_vocab is sparse
+                byte_to_ids[byte_sequence].append(token_id)
+
+        duplicates = {
+            byte_seq: ids for byte_seq, ids in byte_to_ids.items() if len(ids) > 1
+        }
+        if duplicates:
+            error_message = "Duplicate byte sequences found in vocabulary from decode_vocab. Cannot create unique byte->ID mapping (_encode). Duplicates:\n"
+            for byte_seq, ids in duplicates.items():
+                error_message += (
+                    f"  - Bytes {byte_seq!r} correspond to token IDs: {ids}\n"
+                )
+            raise ValueError(error_message)
+
+        _merges = _extract_bpe_merges(tokenizer)
+
+        # Build _encode (bytes -> token_id map) from _decode
+        _encode = {b: i for i, b in enumerate(_decode) if b is not None}
+
+        # Build _encode_byte (single byte -> token_id map)
+        _encode_byte = [None] * 256
+        for i in range(256):
+            byte_val = bytes([i])
+            if byte_val in _encode:
+                _encode_byte[i] = _encode[byte_val]
+
+        eos_token_ids = llm.token_maps.eos_idxs  # Assume this is already a list or set
+        if not eos_token_ids:
+            warnings.warn(
```

I'd raise an error here if this actually has the effect of never allowing termination.